modules/e2ee/E2EEncryption.js

```diff
@@ -53,32 +154,89 @@ export class E2EEncryption {
     }
 
     /**
-     * Setup E2EE on the new track that has been added to the conference, apply it on all the open peerconnections.
-     * @param {JitsiLocalTrack} track - the new track that's being added to the conference.
+     * Advances (using ratcheting) the current key whern a new participant joins the conference.
      * @private
      */
-    _onLocalTrackAdded(track) {
-        for (const session of this.conference._getMediaSessions()) {
-            this._setupSenderE2EEForTrack(session, track);
+    _onParticipantJoined(id) {
+        logger.debug(`Participant ${id} joined`);
+
+        if (this._conferenceJoined && this._enabled) {
+            this._ratchetKey();
+        }
+    }
+
+    /**
+     * Rotates the current key when a participant leaves the conference.
+     * @private
+     */
+    _onParticipantLeft(id) {
+        logger.debug(`Participant ${id} left`);
+
+        this._e2eeCtx.cleanup(id);
+
+        if (this._enabled) {
+            this._rotateKey();
         }
     }
 
     /**
-     * Sets the key to be used for End-To-End encryption.
+     * Event posted when the E2EE signalling channel has been establioshed with the given participant.
+     * @private
+     */
+    _onParticipantE2EEChannelReady(id) {
+        logger.debug(`E2EE channel with participant ${id} is ready`);
+    }
+
+    /**
+     * Handles an update in a participant's key.
      *
-     * @param {string} key - the key to be used.
-     * @returns {void}
+     * @param {string} id - The participant ID.
+     * @param {Uint8Array | boolean} key - The new key for the participant.
+     * @private
+     */
+    _onParticipantKeyUpdated(id, key) {
+        logger.debug(`Participant ${id} updated their key`);
+
+        this._e2eeCtx.setKey(id, key);
+    }
+
+    /**
+     * Advances the current key by using ratcheting.
+     * TODO: not yet implemented, we are just rotating the key at the moment,
+     * which is a heavier operation.
+     * TODO: debounce.
+     *
+     * @private
+     */
+    async _ratchetKey() {
+        return this._rotateKey();
+    }
+
+    /**
+     * Rotates the local key. Rotating the key implies creating a new one, then distributing it
+     * to all participants and once they all received it, start using it.
+     * TODO: debounce.
+     *
+     * @private
      */
-    setKey(key) {
-        this._e2eeCtx.setKey(key);
+    async _rotateKey() {
+        const key = this._generateKey();
+
+        await this._olmAdapter.updateKey(key);
+
+        this._e2eeCtx.setKey(this.conference.myUserId(), key);
```

Comment:
design question: this means we have the keys in userland js, not in the worker?

Reply:
Yes. We generate it outside of the worker and then send it to the worker. In addition, we pass it on the olm adapter, which sends it to everyone else, and keeps it around to send it in ACKs. Now, we can probably get rid of this, since we regenerate the key when a new participant joins anyway, and once we add ratcheting, we'll just send a message to the worker to ratchet.